src/ode_trainer.py

```diff
@@ -0,0 +1,204 @@
+                                eLow=SED_ENERGY_MIN, eHigh=SED_ENERGY_MAX, N=2000,  logGrid=True,
+                                starMassMin=5, starMassMax=500, imfBins=100, imfIndex=2.35, fEsc=starsEscFrac[i][0],
+                                alpha=qsoAlpha[i][0], qsoEfficiency=qsoEfficiency[i][0],
+                                targetSourceAge=sourceAge[i][0])
+        sed_vector.append(energies)
+    sed_vector = np.asarray(sed_vector)
+
+    # sample SED vector
+    x_H_II = np.random.uniform(ps_ode[0][0], ps_ode[0][1], size=(train_set_size, 1))
+    x_He_II = np.random.uniform(ps_ode[1][0], ps_ode[1][1], size=(train_set_size, 1))
+    x_He_III = np.random.uniform(ps_ode[2][0], ps_ode[2][1], size=(train_set_size, 1))
+    T = np.random.uniform(ps_ode[3][0], ps_ode[3][1], size=(train_set_size, 1))
+    tau = np.random.uniform(ps_ode[4][0], ps_ode[4][1], size=(train_set_size, 1))
+    time = np.random.uniform(ps_ode[5][0], ps_ode[5][1], size=(train_set_size, 1))
+
+    state_vector = np.concatenate((x_H_II, x_He_II, x_He_III, T, tau, time), axis=1)
+
+    # sample target labels
+    u_actual = np.zeros((train_set_size, 1))
+
+    return sed_vector, state_vector, u_actual
+
+
+
+# -----------------------------------------------------------------
+#  Main
+# -----------------------------------------------------------------
+def main(config):
+    # -----------------------------------------------------------------
+    # create unique output path and run directories, save config
+    # -----------------------------------------------------------------
+    run_id = 'run_' + utils_get_current_timestamp()
+    config.out_dir = os.path.join(config.out_dir, run_id)
+
+    utils_create_output_dirs([config.out_dir])
+    # utils_create_run_directories(config.out_dir, DATA_PRODUCTS_DIR, PLOT_DIR)
+    utils_save_config_to_log(config)
+    utils_save_config_to_file(config)
+
+    # data_products_path = os.path.join(config.out_dir, DATA_PRODUCTS_DIR)
+    # plot_path = os.path.join(config.out_dir, PLOT_DIR)
+
+    # initialise the model
+    u_approximation = MLP1(config)
+
+    if cuda:
+        model.cuda()
+
+    # initialise the ODE equation
+    ode_equation = ODE()
+
+    # -----------------------------------------------------------------
+    # Optimizers
+    # -----------------------------------------------------------------
+    optimizer = torch.optim.Adam(
+        u_approximation.parameters(),
+        lr=config.lr,
+        betas=(config.b1, config.b2)
+    )
+
+    # -----------------------------------------------------------------
+    # book keeping arrays
+    # -----------------------------------------------------------------
+    train_loss_array = np.empty(0)
+    val_loss_mse_array = np.empty(0)
+    val_loss_dtw_array = np.empty(0)
+
+    print("\033[96m\033[1m\nTraining starts now\033[0m")
+    for epoch in range(1, config.n_epochs + 1):
+
+        # TODO: look for boundary conditions???
```

Comment:
Some boundary conditions are:  for r=0, x_H_II = x_He_III = 1 and x_H_I = x_He_I = x_He_II = 0 Meaning everything is fully ionised.